### PR TITLE
Code refactoring: Create new deployment backend class

### PR DESCRIPTION
**Pull request by noliveleger**
## Notes

Part II of #5036. 
Create a new deployment class: `OpenRosaDeploymentBackend` which becomes the default one. 

:warning: even if the tests are passing, it contains some bugs with partial permissions caught but tests in #5056

## Related issues

#5036

### CODE REVIEW

File: kpi/deployment_backends/openrosa_backend.py [hunk 1]
```diff
@@ -0,0 +1,1550 @@
...
+        # Remove KoboCAT orphan files previously uploaded through KPI
+        for metadata_file in metadata_files.values():
+            if metadata_file['from_kpi']:
+                self._delete_openrosa_metadata(metadata_file)
+
+    @property
+    def xform(self):
+
+        if self._xform is not None:
+            return self._xform
```
Comment (jnm) [inline kpi/deployment_backends/openrosa_backend.py:1299]: :hugs:

File: kpi/deployment_backends/mock_backend.py [hunk 2]
```diff
@@ -105,7 +105,7 @@ def generate_uuid_for_form():
     def form_uuid(self):
         return 'formhub-uuid'  # to match existing tests
 
-    def nlp_tracking_data(self, start_date=None):
+    def nlp_tracking_data(asset_ids=None):
```
Comment (jnm) [inline kpi/deployment_backends/mock_backend.py:108]: is this a mistake? i see `start_date` still referenced in the method (and `asset_ids` does not appear)

**Review by jnm — CHANGES_REQUESTED** [review]

File: kobo/apps/trash_bin/utils.py [hunk 3]
```diff
@@ -315,8 +310,8 @@ def _delete_submissions(request_author: settings.AUTH_USER_MODEL, asset: 'kpi.As
         submission_ids = []
         for submission in submissions:
             audit_logs.append(AuditLog(
-                app_label=app_label,
-                model_name=model_name,
+                app_label='logger',
+                model_name='instance',
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply kobo/apps/trash_bin/utils.py:314]: What do you suggest? Existing logs are already using `logger` and `instance`

File: kobo/settings/base.py [hunk 4]
```diff
@@ -1707,23 +1708,11 @@ def dj_stripe_request_callback_method():
...
-        'uid': 'validation_status_on_hold',
-        'color': '#0000ff',
-        'label': 'On Hold'
-    },
-]
+DEFAULT_VALIDATION_STATUSES = {
+    'validation_status_not_approved': 'Not Approved',
+    'validation_status_approved': 'Approved',
+    'validation_status_on_hold': 'On Hold',
+}
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply kobo/settings/base.py:1715]: Indeed, the front end never used the color provided by the back end. 
When we wrote this part of code (the validation status feature), the idea behind this was to have a specific setting which could override the setting. Each time I looked at it, I found it pretty complicated to understand. If we were to implement that on the front end, we would to overwrite on the fly the colors for existing data (because colors are hard-coded in Mongo).
Long story short, I'll fix the API docstring and formpack test fixtures.

File: kpi/tasks.py [hunk 5]
```diff
@@ -73,7 +76,14 @@ def sync_kobocat_xforms(
 
 @celery_app.task
 def sync_media_files(asset_uid):
-    asset = Asset.objects.get(uid=asset_uid)
+    asset = Asset.objects.defer('content').get(uid=asset_uid)
+    if not asset.has_deployment:
+        # 🙈 Race condition: Celery task starts too fast and does not see
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply kpi/tasks.py:81]: Probably. I would like to get rid of my hacky solution. 
I could add a `TODO`.

File: kpi/deployment_backends/openrosa_backend.py [hunk 6]
```diff
@@ -0,0 +1,1550 @@
...
+            if element is not None:
+                element.text = date_formatted
+        # Rely on `meta/instanceID` being present. If it's absent, something is
+        # fishy enough to warrant raising an exception instead of continuing
+        # silently
+        xml_parsed.find(self.SUBMISSION_CURRENT_UUID_XPATH).text = (
+            uuid_formatted
+        )
+
+        safe_create_instance(
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply kpi/deployment_backends/openrosa_backend.py:322]: You are totally right. I think I will leaving in a world where 500 errors do not exist 🤦

File: kpi/deployment_backends/openrosa_backend.py [hunk 7]
```diff
@@ -0,0 +1,1550 @@
...
+                title=self.asset.name,
+                has_kpi_hooks=self.asset.has_active_hooks,
+            )
+            self.xform.downloadable = active
+            self.xform.title = self.asset.name
+            self.xform.has_kpi_hooks = self.asset.has_active_hooks
+
+            publish_xls_form(xlsx_file, self.asset.owner, self.xform.id_string)
+
+        # Do not call save it, asset (and its deployment) is saved right
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply kpi/deployment_backends/openrosa_backend.py:848]: Indeed!

File: kpi/deployment_backends/openrosa_backend.py [hunk 8]
```diff
@@ -0,0 +1,1550 @@
...
+        self.store_data(
+            {
+                'backend': 'openrosa',
+                'active': active,
+                'backend_response': {
+                    'formid': self.xform.pk,
+                    'uuid': self.xform.uuid,
+                    'id_string': self.xform.id_string,
+                    'kpi_asset_uid': self.asset.uid,
+                },
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply kpi/deployment_backends/openrosa_backend.py:859]: It is! From my tests, I created a bunch of project with KobocatDeploymentBackend class, then switched them. I could redeploy them without any issues. TBH, I don't remember if I could redeployed with KobocatDeploymentBackend after switching them but I think I could.

File: kpi/deployment_backends/base_backend.py [hunk 9]
```diff
@@ -172,18 +172,19 @@ def bulk_update_submissions(
             for path, value in update_data.items():
                 edit_submission_xml(xml_parsed, path, value)
 
-            kc_response = self.store_submission(
-                user, xml_tostring(xml_parsed), _uuid
+            backend_response = self.store_submission(
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply kpi/deployment_backends/base_backend.py:175]: Actually it is handled in [`prepare_bulk_update_response`](https://github.com/kobotoolbox/kpi/blob/06db7915939859cc42e3c12dc0714acc3efe4b25/kpi/deployment_backends/openrosa_backend.py#L923). 
I've tried to keep the same logic we have in KobocatDeploymentBackend (i.e. validating response from KC in `prepare_bulk_update_response`). This could be refactored too in #5056 too if we want to catch errors earlier.

File: kpi/deployment_backends/mock_backend.py (hunk 2 above)
Comment (noliveleger) [reply kpi/deployment_backends/mock_backend.py:108]: It is indeed. It should be 

```python 
@staticmethod
def nlp_tracking_data(asset_ids=None, start_date=None):
    ...
```

but this method is gone in #5056. Fixing it here will cause conflicts when this branch is merged in #5056. Is it really worth it? 


Out of scope: It really means the tests were not covering this part.

File: kpi/deployment_backends/openrosa_backend.py [hunk 10]
```diff
@@ -0,0 +1,1550 @@
...
+        except Instance.DoesNotExist:
+            return {
+                'content_type': 'application/json',
+                'status': status.HTTP_404_NOT_FOUND,
+                'data': {
+                    'detail': f'No submission found with ID: {submission_id}'
+                }
+            }
+
+        if method == 'DELETE':
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply kpi/deployment_backends/openrosa_backend.py:1077]: Actually, I used typing annotation to specify what would be the expected VALUES. What do you think? 

``` 
    def set_validation_status(
        self,
        submission_id: int,
        user: settings.AUTH_USER_MODEL,
        data: dict,
        method: str = Literal['DELETE', 'PATCH'],
    ) -> dict:
```

File: kpi/deployment_backends/openrosa_backend.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (noliveleger) [reply kpi/deployment_backends/openrosa_backend.py]: I have added "TODOs" and I will create a PR on top of #5056 to fix all `safe_create_instance` calls.

File: kpi/deployment_backends/openrosa_backend.py [hunk 11]
```diff
@@ -0,0 +1,1550 @@
...
+        )
+
+        if not (
+            xform
+            and xform.user.username == self.asset.owner.username
+            and xform.id_string == self.xform_id_string
+        ):
+            raise InvalidXFormException(
+                'Deployment links to an unexpected KoboCAT XForm'
+            )
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply kpi/deployment_backends/openrosa_backend.py:1330]: As per discussed. We only have to support `Asset` object deletion even if related `XForm` does not exist.

[the remark this answers is not in the document]
Comment (noliveleger) [reply kpi/deployment_backends/openrosa_backend.py:1330]: Fixed in #5075

File: kobo/apps/trash_bin/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jnm) [reply kobo/apps/trash_bin/utils.py:314]: We discussed static vs. dynamic, and it seemed like a fairly arbitrary source. Static avoids an import

File: kpi/deployment_backends/openrosa_backend.py [hunk 12]
```diff
@@ -0,0 +1,1550 @@
...
+
+        safe_create_instance(
+            username=user.username,
+            xml_file=ContentFile(xml_tostring(xml_parsed)),
+            media_files=attachments,
+            uuid=_uuid,
+            request=request,
+        )
+        return self._rewrite_json_attachment_urls(
+            next(self.get_submissions(user, query={'_uuid': _uuid})), request
```
[the remark this answers is not in the document]
Comment (jnm) [reply kpi/deployment_backends/openrosa_backend.py]: will be handled in a future PR added to the #5036 series

File: kpi/deployment_backends/openrosa_backend.py [hunk 13]
```diff
@@ -0,0 +1,1550 @@
...
+        xml_submission_file.seek(0)
+
+        # Retrieve only File objects to pass to `safe_create_instance`
+        # TODO remove those files as soon as the view sends request.FILES directly
+        #   See TODO in kpi/views/v2/asset_snapshot.py::submission
+        media_files = (
+            media_file for media_file in attachments.values()
+        )
+
+        safe_create_instance(
```
[the remark this answers is not in the document]
Comment (jnm) [reply kpi/deployment_backends/openrosa_backend.py:403]: > I have added "TODOs" and I will create a PR on top of #5056 to fix all `safe_create_instance` calls.

https://github.com/kobotoolbox/kpi/pull/5038#discussion_r1727117366

File: kpi/deployment_backends/openrosa_backend.py [hunk 14]
```diff
@@ -0,0 +1,1550 @@
...
+            username=user.username,
+            xml_file=xml_submission_file,
+            media_files=media_files,
+            request=request,
+        )
+
+        return {
+            'headers': {},
+            'content_type': 'text/xml; charset=utf-8',
+            'status': status.HTTP_201_CREATED,
```
[the remark this answers is not in the document]
Comment (jnm) [reply kpi/deployment_backends/openrosa_backend.py:413]: > I have added "TODOs" and I will create a PR on top of #5056 to fix all `safe_create_instance` calls.

https://github.com/kobotoolbox/kpi/pull/5038#discussion_r1727117366

File: kpi/deployment_backends/openrosa_backend.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jnm) [reply kpi/deployment_backends/openrosa_backend.py:1082]: looks good! i was mainly worried about non-developers hitting the API with invalid verbs, but as you pointed out, that's already handled

File: kpi/deployment_backends/mock_backend.py (hunk 2 above)
Comment (jnm) [reply kpi/deployment_backends/mock_backend.py:108]: if it's gone in #5056, let's not worry about it :)
